### ISSUE TITLE
Embedded discriminators + updateOne + $set = no-go

### ISSUE THREAD

**Issue by davebaol** [Automattic/mongoose#5130]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
As the title says `embedded discriminators + updateOne + $set = no-go` where no-go means that the discriminator key and the properties of the specialized subdocument are lost.

**If the current behavior is a bug, please provide the steps to reproduce.**
The test below is a modified version of the test at commit https://github.com/Automattic/mongoose/commit/7d6c2594af889e6e2a47e8ae2c217171af778bf1

````javascript
  it('embedded discriminators with $set (gh-5130)', function(done) {
    var eventSchema = new Schema({ message: String },
      { discriminatorKey: 'kind' });
    var batchSchema = new Schema({ events: [eventSchema] });
    var docArray = batchSchema.path('events');

    var Clicked = docArray.discriminator('Clicked', new Schema({
      element: {
        type: String,
        required: true
      }
    }));

    var Purchased = docArray.discriminator('Purchased', new Schema({
      product: {
        type: String,
        required: true
      }
    }));

    var Batch = mongoose.model('gh5130', batchSchema);

    var batch = {
      events: [
        { kind: 'Clicked', element: '#hero' }
      ]
    };

    Batch.create(batch).
      then(function(doc) {
        assert.equal(doc.events.length, 1);
        return Batch.updateOne({ _id: doc._id, 'events._id': doc.events[0]._id }, {
            $set: {
                'events.$':  { message: 'updated', kind: 'Clicked', element: '#hero2' } 
              }
            })
            .then(function() {
               return doc;
            });
      }).
      then(function(doc) {
        return Batch.findOne({ _id: doc._id });
      }).
      then(function(doc) {
        assert.equal(doc.events.length, 1);
        assert.equal(doc.events[0].message, 'updated');
        assert.equal(doc.events[0].element, '#hero2');    // <-- test failed
        assert.equal(doc.events[0].kind, 'Clicked');      // <-- test failed 
        done();
      }).
      catch(done);
  });
````

**What is the expected behavior?**
Properties `element` and `kind` should be persisted.

**Please mention your node.js, mongoose and MongoDB version.**
- node 6.9.4
- mongodb 3.4.2
- mongoose 4.9.2

**Comment by davebaol** [comment]
Just tested the brand-new version 4.9.3. 
The test above still fails :)

**Comment by varunjayaraman** [comment]
@davebaol thanks, this looks like a bug

**Comment by vkarpov15** [comment]
Fixed, thanks for the detailed repro :+1: Fix will be in 4.9.4, ETA April 9.

**Comment by davebaol** [comment]
@vkarpov15 @varunjayaraman 
Thank you, guys.
As usual, you do real good work on this project. :clap:
